src/backend/joanie/payment/backends/base.py

```diff
@@ -168,13 +172,65 @@ def _send_mail_payment_installment_success(
                 to_user_email=order.owner.email,
             )
 
-    @staticmethod
-    def _do_on_payment_failure(order, installment_id):
+    @classmethod
+    def _send_mail_refused_debit(cls, order, installment_id):
+        """
+        Prepare mail context when debit has been refused for an installment in the
+        the current language of the user.
+        """
+        with override(order.owner.language):
+            title = order.product.safe_translation_getter(
+                "title", language_code=order.owner.language
+            )
+            amount = Money(
+                next(
+                    installment["amount"]
+                    for installment in order.payment_schedule
+                    if installment["id"] == installment_id
+                ),
+                currency=settings.DEFAULT_CURRENCY,
+            )
```

Comment:
IMO you should handle properly the case where the installment id is not valid and raise a `KeyError`

Reply:
I think it will repeat what does another [method](https://github.com/openfun/joanie/blob/5f7d962d577bfb63684d40452483b089e1b9eac9/src/backend/joanie/core/models/products.py#L1135) here in the Order Model. If the `installment_id` is not found, it raises a `ValueError` in that other [method](https://github.com/openfun/joanie/blob/5f7d962d577bfb63684d40452483b089e1b9eac9/src/backend/joanie/core/models/products.py#L1112) called in the cycle.

The entry point, the method `_do_on_payment_failure` calls the first method mentioned above :
```python
    @classmethod
    def _do_on_payment_failure(cls, order, installment_id):
        """
        Generic actions triggered when a failed payment has been received.
        Mark the invoice as pending.
        """
        order.set_installment_refused(installment_id)
        cls._send_mail_refused_debit(order, installment_id)
```

While using `next()` built-in object from Python, it will raise a `StopIteration` if it cannot find the installment["id"] into the payment schedule of the order. 
Else, I can also make a `KeyError` if the `amount` key is missing from the payment schedule.

Let me know what suits you best.
For now I made this locally : 
```
            try:
                amount = Money(
                    next(
                        installment["amount"]
                        for installment in order.payment_schedule
                        if installment["id"] == installment_id
                    ),
                    currency=settings.DEFAULT_CURRENCY,
                )
            except StopIteration as exception:
                raise ValueError(
                    f"Payment Base Backend: {installment_id} not found!"
                ) from exception
            except KeyError as exception:
                raise ValueError(
                    "Payment Base Backend: 'amount' key is not found in payment schedule with"
                    f"passed 'installment_id' {installment_id} value."
                ) from exception
```

Reply:
Looks good to manage the `StopIteration` exception. On the other hand, IMO you should remove the `KeyError` exception management. We can assume that the order.payment_schedule is always well formed.